### PR TITLE
Advent of Code Migration from Sir Lancebot

### PR DESCRIPTION
**Pull request by janine9vn**
This is a migration of the Advent of Code cog, mostly as-is, from Sir Lancebot to Sir Robin. This is as close to a straight migration as I could manage. Any changes or improvements we want will be made in subsequent PRs once this is merged in.

### How do I review this?
I've split this up into 4 commits, grouped by collections of files/changes:
- Porting over the necessary `utils/` files
- Updating the `constants.py` file
- A quick port of the invoke help command thingy
- Migration of the Advent of Code cog itself

The commits have information about what is not quite the straight port. Spoiler alert: it's some import statements and parts of the Redis Cache. Everything else is a straight copy/paste or drag/drop.

If you'd like to test this with minimal effort, I can provide most of my `.env` file that includes the leaderboard session cookie information. I can also provide you access to my test server where it *should* have the roles and constants configured for my `.env` file.

### CODE REVIEW

File: bot/bot.py [hunk 1]
```diff
@@ -70,3 +72,11 @@ async def send_log(self, title: str, details: str = None, *, icon: str = None) -
         embed.set_author(name=title, icon_url=icon)
 
         await devlog.send(embed=embed)
+
+    async def invoke_help_command(self, ctx: commands.Context) -> None:
+        """Invoke the help command or default help command if help extensions is not loaded."""
+        if "bot.exts.core.help" in ctx.bot.extensions:
```
Comment (minalike) [inline bot/bot.py:77]: Is this what the help command is supposed to look like? This looks like the vanilla.

image

Comment (camcaswell) [reply bot/bot.py:77]: I'd rather have it merged than worry about the help text for now

Comment (minalike) [reply bot/bot.py:77]: I misunderstood what the help invoke function is about. It works (ex: &help aoc x)

**Review by minalike — APPROVED** [review]

File: bot/exts/advent_of_code/_helpers.py [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+        # The Advent of Code website redirects silently with a 200 response if a
+        # session cookie has expired, is invalid, or was not provided.
+        if str(resp.url) != url:
+            log.error(f"Fetching leaderboard `{board}` failed! Check the session cookie.")
+            raise UnexpectedRedirect(f"redirected unexpectedly to {resp.url} for board `{board}`")
+
+        # Every status other than `200` is unexpected, not only 400+
+        if not resp.status == 200:
+            log.error(f"Unexpected response `{resp.status}` while fetching leaderboard `{board}`")
+            log.error(resp)
```
Comment (MarkKoz) [inline bot/exts/advent_of_code/_helpers.py:239]: Was this accidentally left in while locally debugging?

File: bot/exts/advent_of_code/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+        }
+
+        # Store the new values in Redis
+        await _caches.leaderboard_cache.update(cached_leaderboard)
+
+        # Set an expiry on the leaderboard RedisCache
+        await bot.instance.redis_session.client.expire(
+            _caches.leaderboard_cache.namespace,
+            AdventOfCode.leaderboard_cache_expiry_seconds
+        )
```
Comment (MarkKoz) [inline bot/exts/advent_of_code/_helpers.py:357]: Basically this is the only other change and it looks correct 👍

**Review by MarkKoz — APPROVED** [review]
Can someone make time to figure out which utilities are common and create an issue for them to be migrated to bot-core? The copy-pasting is fine for now given the urgency.

Will the AoC extension on the other bot be disabled for the time being, or is there a PR soon to remove that code?

**Comment by MarkKoz** [comment]
I did not intend for my comments to block the merge (given that they were left with an approval). I don't know why GH is blocking. Feel free to resolve them whenever you want to merge this. I assume you're still waiting for env vars to be set in prod?

File: bot/exts/advent_of_code/_helpers.py (hunk 2 above)
Comment (janine9vn) [reply bot/exts/advent_of_code/_helpers.py:239]: Yes, but also I do want to keep it in for a little bit. I had an issue with a leaderboard and the status alone wasn't enough info to know what happened.

File: bot/exts/advent_of_code/_helpers.py (hunk 3 above)
Comment (janine9vn) [reply bot/exts/advent_of_code/_helpers.py:357]: Praise be to Scale who showed me how the new async-rediscache stuff works

**Pull request merged**